subnet routes (fixed, completed)

BenjaminLudwigSAP opened the pull request:
This fixes and completes #196 and #202.
This will be followed up by - but is independent of - refactoring PR #205.
Closes #207

notandy committed:
Interface route for subnets without self-ip

This PR introduces l2 generation and update for interface routes for
subnets of a network that have no self-ip.
This allows having pool members from foreign subnets without the need to route
via default gw and allow setups like no-snat.

BenjaminLudwigSAP committed:
[l2] Full sync: Clean up orphaned subnet routes

BenjaminLudwigSAP committed:
Fix subnet routes task: Fails when no routes present on device

BenjaminLudwigSAP committed:
[l2] Fix subnet route provisioning

- Provision subnet routes to Common tenant, because
  - Tenant deletion happens before L2 cleanup and fails when routes
    are still present in the tenant
  - A small caveat is that we can't filter by regex or wildcard when
    fetching routes from the BigIP via iControl REST, so we need to
    filter after fetching.

- Rename methods to something with 'subnet routes' instead of 'static
  routes'. The routes are static insofar as they are statically
  configured, but so are other routes provisioned by the F5 provider
  driver, so 'static' is not needed.

- A separate CleanupSubnetRoutes task is needed since SyncSubnetRoutes
  does not know whether the network is to be deleted (i. e. whether it
  was called from remove_l2) and happily creates routes for all
  subnets of a network at network tenant deletion.

- Fix: SyncSubnetRoutes creates only one single subnet route due to
  unconditional return statement in the for loop.

- Fix: Missing RaisesIControlRestError decorator on
  CleanupDefaultRoute execute method

- Simplify subnet route sync logic

TODO:

- Remove orphaned subnet routes in L2SyncManager.full_sync

- SelfIPs and subnet routes are mutually exclusive. Their creation and
  deletion has to be coordinated, respectively

BenjaminLudwigSAP committed:
[unittest] test_SyncSubnetRoutes: Check subnet route name

notandy commented on octavia_f5/controller/worker/tasks/f5_tasks.py:
```diff
@@ -297,6 +368,33 @@ def execute(self, network: f5_network_models.Network,
+        for existing_route in existing_routes:
+            existing_route_name = existing_route['name']
+            if existing_route_name.startswith(subnet_route_network_prefix):
+                res = bigip.delete(path=existing_route['path'])
```
same as SyncSubnetRoutes, there is no path element.

notandy commented on octavia_f5/controller/worker/tasks/f5_tasks.py:
```diff
@@ -272,10 +278,75 @@ def execute(self, bigip: bigip_restclient.BigIPRestClient,
+
+            # Delete unneeded route
+            if delete_device_route:
+                res = bigip.delete(path=existing_route['path'])
```
there is no path element in the bigip response:
```
2023-03-03 12:17:50 21842 ERROR octavia_f5.controller.worker.l2_sync_manager Traceback (most recent call last):
2023-03-03 12:17:50 21842 ERROR octavia_f5.controller.worker.l2_sync_manager   File "/Users/d072895/.virtualenvs/yoga_python3.8.10/lib/python3.8/site-packages/taskflow/engines/action_engine/executor.py", line 53, in _execute_task
2023-03-03 12:17:50 21842 ERROR octavia_f5.controller.worker.l2_sync_manager     result = task.execute(**arguments)
2023-03-03 12:17:50 21842 ERROR octavia_f5.controller.worker.l2_sync_manager   File "/Users/d072895/.pyenv/versions/3.8.10/lib/python3.8/contextlib.py", line 75, in inner
2023-03-03 12:17:50 21842 ERROR octavia_f5.controller.worker.l2_sync_manager     return func(*args, **kwds)
2023-03-03 12:17:50 21842 ERROR octavia_f5.controller.worker.l2_sync_manager   File "/Users/d072895/Workspace/yoga/octavia-f5-provider-driver/octavia_f5/controller/worker/tasks/f5_tasks.py", line 327, in execute
2023-03-03 12:17:50 21842 ERROR octavia_f5.controller.worker.l2_sync_manager     res = bigip.delete(path=existing_route['path'])
2023-03-03 12:17:50 21842 ERROR octavia_f5.controller.worker.l2_sync_manager KeyError: 'path'
```

I assume you meant fullPath - but it won't work since it needs the full api path. I would construct it from it's name: `path=f"/mgmt/tm/net/route/~Common~{existing_route['name']}"`

BenjaminLudwigSAP replied:
Interesting. [I didn't change that since the first PR](https://github.com/sapcc/octavia-f5-provider-driver/pull/196/files#diff-534649279f55ec529e3899f23bc1a7526a18aaf4fb399defafbaa542c12acc9aR309), so the error was there all along.

notandy commented on octavia_f5/controller/worker/tasks/f5_tasks.py:
```diff
@@ -272,10 +278,75 @@ def execute(self, bigip: bigip_restclient.BigIPRestClient,
+
+                # if the route is a needed subnet route, don't touch it
+                if existing_route_name == get_subnet_route_name(network.id, subnet_that_needs_route):
+                    subnets_that_need_routes.remove(subnet_that_needs_route)
```
This is a very bad style. You don't want to modify the array you're currently iterating.
You can easily rewrite it by building a list of subnet-route-names and iterate through the disjoint elements.

BenjaminLudwigSAP replied:
Good find.
Modifying the array we're currently iterating over is only risky as soon as we enter the next iteration. Right now I'm gravitating to simply putting a `break` statement after line 323 to make sure this is the last iteration, since at this point we're already sure we can ignore the subnet.
Creating a whole new array just for iteration is a bit too heavy in this specific instance, as compared to a `break`.

BenjaminLudwigSAP committed:
[l2] Refactor subnet route provisioning

Ignore existing and needed route immediately without iterating over
the rest of the needed subnet routes.
This does not change behavior as long as each subnet ID appears only
once in the list of subnets, which we know is the case.

BenjaminLudwigSAP committed:
[l2] Fix subnet route deletions

existing_route does not have a 'path' key. This error was introduced
in #196

notandy committed:
[l2] don't mutate array while iterating over it

Same functionalty as before, but without changing the array that is
being iterated and a bit more compact.

notandy commented:
I've reproduced that the feature is suffering from the bug, that happens when adding a load-balancer to a subnet that already got a static subnet route. But so far it looks good.

BenjaminLudwigSAP committed:
[l2] Fix: Subnet routes conflict with SelfIPs

- L2SyncManager:

  - Rename sync_l2_selfips_flow to
    sync_l2_selfips_and_subnet_routes_flow.
    I may in a future PR rename this again to something that makes the
    dichotomy more clear between provisioning all l2 objects and only
    SelfIPs and subnet routes. Maybe something along the lines of
    sync_l2_network_scope vs. sync_l2_subnet_scope, we'll see.

  - If there are no SelfIPs to provision (i.e. selfips is empty) we
    don't return immediately anymore, but instead provision subnet
    routes for all subnets. This gives us the least surprising behavior
    in the event that no SelfIP ports were created.

  - _do_sync_l2_selfips_and_subnet_routes_flow:
    Don't calculate SelfIPs to add/remove here. That calculation has
    to happen in the cleanup_selfips_and_subnet_routes and
    ensure_selfips_and_subnet_routes flows, respectively. This is so
    that selection of which subnet routes to delete can happen in the
    CleanupSubnetRoutes task. This avoids code duplication because
    that task is also called from remove_l2, but to delete ALL subnet
    routes of a network.

- F5Flows: Split up sync_l2_selfips into
  cleanup_selfips_and_subnet_routes for deletion and
  ensure_selfips_and_subnet_routes for creation

- F5Tasks:

  - Rename SyncSubnetRoutes to EnsureSubnetRoutes

  - Separate subnet_in_selfips sub-function out of EnsureSubnetRoutes
    to reuse it in both EnsureSubnetRoutes and CleanupSubnetRoutes

  - CleanupSubnetRoutes: Clean up either all subnet routes or only the
    unneeded ones. The former case is used by F5Flows.remove_l2

BenjaminLudwigSAP committed:
[l2] Fix full_sync: Needed subnet routes are deleted

- Calculation of which subnet routes are needed was wrong: All subnets
  of all networks were considered. This means that orphaned subnet
  routes of still existing subnet routes were not deleted, even if the
  subnet doesn't have LBs anymore.

- continue statement was in nested for-loop. This means that no subnet
  routes were really skipped, the for-loop did nothing at all.
  This in turn means that all subnet routes were deleted at this point
  during full_sync and then recreated later when full_sync executes
  ensure_l2_flow.

notandy commented on octavia_f5/controller/worker/l2_sync_manager.py:
```diff
@@ -250,20 +251,17 @@ def sync_l2_selfips_flow(self, selfips: [network_models.Port], network_id: str,
             except Exception as e:
-                self._metric_failed_futures.labels(bigip.hostname, 'sync_l2_selfips_flow').inc()
-                LOG.error("Failed running sync_l2_selfips_flow for host %s: %s", bigip.hostname, e)
+                self._metric_failed_futures.labels(bigip.hostname, 'sync_l2_selfips_and_subnet_routes_flow').inc()
+                LOG.error("Failed running sync_l2_selfips_and_subnet_routes_flow for host %s: %s", bigip.hostname, e)
```
If changing metric names keep in mind changing grafana dashboard.

BenjaminLudwigSAP replied:
If I am not mistaken there is nothing to change on the grafana dashboards. `sync_l2_selfips_and_subnet_routes_flow` is not a metric name, but a value of the `task` label, and it's not explicitly used anywhere on the Octavia grafana dashboard. The Octavia grafana dashboard only includes a `by (...) task` clause.

notandy commented on octavia_f5/controller/worker/flows/f5_flows.py:
```diff
@@ -49,37 +52,66 @@ def remove_l2(self, selfips: [str]) -> flow.Flow:
+            ensure_selfip = f5_tasks.EnsureSelfIP(name=f"ensure-selfip-{selfip.id}", inject={'port': selfip})
+            ensure_selfips_and_subnet_routes_flow.add(ensure_selfip)
+
+        # removal of subnet routes
```
you are not removing subnet routes here

notandy commented on octavia_f5/tests/unit/controller/worker/tasks/test_f5_tasks.py:
```diff
@@ -199,7 +199,7 @@ def test_SyncSubnetRoutes(self, mock_get_subnet):
         )
 
-        engines.run(f5_tasks.SyncSubnetRoutes(),
+        engines.run(f5_tasks.EnsureSubnetRoutes(),
```
how about a test case for cleanup subnet routes and sync_l2_selfips_and_subnet_routes_flow?

BenjaminLudwigSAP replied:
Will do in a seperate PR.

notandy reviewed (approved):
I've added minor comments, but in general it seems to work fine.

Afaik, `_do_sync_l2_selfips_and_subnet_routes_flow` can be completely embedded (and have therefor less boilerplate) in `ensure_selfips_and_subnet_routes` and is not reused elsewhere, but it's fine to do it another time.

BenjaminLudwigSAP commented:
I've opened #211 for future L2 refactoring.
